src/minimega/router_cli.go

```diff
@@ -0,0 +1,283 @@
+			"clear router <vm> <dns,>",
+			"clear router <vm> <dns,> <ip>",
+			"clear router <vm> <ra,>",
+			"clear router <vm> <ra,> <subnet>",
+			"clear router <vm> <route,>",
+			"clear router <vm> <route,> <static,>",
+			"clear router <vm> <route,> <static,> <network>",
+			"clear router <vm> <route,> <ospf,>",
+			"clear router <vm> <route,> <ospf,> <area>",
+			"clear router <vm> <route,> <ospf,> <area> <network>",
+		},
+		Call: wrapBroadcastCLI(cliClearRouter),
+	},
+}
+
+func cliRouter(c *minicli.Command, resp *minicli.Response) error {
+	vmName := c.StringArgs["vm"]
+
+	vm := vms.FindVM(vmName)
+	if vm == nil {
+		return fmt.Errorf("no such vm %v", vmName)
+	}
+
+	if vmName != "" && len(c.BoolArgs) == 0 { // a summary of a specific router
+		rtr := FindRouter(vm)
+		if rtr == nil {
+			return fmt.Errorf("vm %v not a router", vmName)
+		}
+		resp.Response = rtr.String()
+	}
+
+	rtr := FindOrCreateRouter(vm)
+
+	if c.BoolArgs["commit"] {
+		return rtr.Commit()
+	} else if c.BoolArgs["log"] {
+		var level string
+		if c.BoolArgs["fatal"] {
+			level = "fatal"
+		} else if c.BoolArgs["error"] {
+			level = "error"
+		} else if c.BoolArgs["warn"] {
+			level = "warn"
+		} else if c.BoolArgs["info"] {
+			level = "info"
+		} else if c.BoolArgs["debug"] {
+			level = "debug"
+		}
+		rtr.LogLevel(level)
+		return nil
+	} else if c.BoolArgs["interface"] {
+		network, err := strconv.Atoi(c.StringArgs["network"])
+		if err != nil {
+			return fmt.Errorf("invalid network: %v : %v", c.StringArgs["network"], err)
+		}
+		ip := c.StringArgs["IPv4/MASK"]
+
+		return rtr.InterfaceAdd(network, ip)
+	} else if c.BoolArgs["dhcp"] {
+		addr := c.StringArgs["listen"]
+
+		if c.BoolArgs["range"] {
+			low := c.StringArgs["low"]
+			high := c.StringArgs["high"]
+			return rtr.DHCPAddRange(addr, low, high)
+		} else if c.BoolArgs["router"] {
+			r := c.StringArgs["router"]
+			return rtr.DHCPAddRouter(addr, r)
+		} else if c.BoolArgs["dns"] {
+			dns := c.StringArgs["address"]
+			return rtr.DHCPAddDNS(addr, dns)
+		} else if c.BoolArgs["static"] {
+			mac := c.StringArgs["mac"]
+			ip := c.StringArgs["ip"]
+			return rtr.DHCPAddStatic(addr, mac, ip)
+		}
+	} else if c.BoolArgs["dns"] {
+		ip := c.StringArgs["ip"]
+		hostname := c.StringArgs["hostname"]
+		rtr.DNSAdd(ip, hostname)
+		return nil
+	} else if c.BoolArgs["ra"] {
+		subnet := c.StringArgs["subnet"]
+		rtr.RADAdd(subnet)
+		return nil
+	} else if c.BoolArgs["route"] {
+		if c.BoolArgs["static"] {
+			network := c.StringArgs["network"]
+			nh := c.StringArgs["next-hop"]
+			rtr.RouteStaticAdd(network, nh)
+			return nil
+		} else if c.BoolArgs["ospf"] {
+			area := c.StringArgs["area"]
+			iface := c.StringArgs["network"]
+			rtr.RouteOSPFAdd(area, iface)
+		}
+	}
+
+	return nil
+}
+
+func cliClearRouter(c *minicli.Command, resp *minicli.Response) error {
+	vmName := c.StringArgs["vm"]
+
+	vm := vms.FindVM(vmName)
+	if vm == nil {
+		return fmt.Errorf("no such vm %v", vmName)
+	}
+	rtr := FindRouter(vm)
```

Comment:
(Somewhat note to self, still figuring out how this all works)

Should we hold `rtr.lock` before clearing things? Each of the `rtr.*Del` methods could grab the lock. Or does `cmdLock` synchronize access to `rtr` -- in which case, do we need `routerLock` and `rtr.lock`?